packages/netlify/src/index.ts

```
@@ -1,2 +1,220 @@
+
+	async function writeRedirects(routes: RouteData[], dir: URL) {
+		const fallback = _config.output === 'static' ? '/.netlify/static' : '/.netlify/functions/ssr';
+		const redirects = createRedirectsFromAstroRoutes({
+			config: _config,
+			dir,
+			routeToDynamicTargetMap: new Map(
+				routes
+					.filter(isStaticRedirect) // all other routes are handled by SSR
+					.map((route) => {
+						// this is needed to support redirects to dynamic routes
+						// on static. not sure why this is needed, but it works.
+						route.distURL ??= route.redirectRoute?.distURL;
+
+						return [route, fallback];
+					})
+			),
+		});
+
+		if (!redirects.empty()) {
+			await appendFile(new URL('_redirects', outDir), '\n' + redirects.print() + '\n');
+		}
+	}
+
+	async function writeSSRFunction() {
+		await writeFile(
+			new URL('./ssr.mjs', ssrOutputDir()),
+			`
+				import createSSRHandler from './entry.mjs';
+				export default createSSRHandler(${JSON.stringify({
+					cacheOnDemandPages: Boolean(integrationConfig?.cacheOnDemandPages),
+				})});
+				export const config = { name: "Astro SSR", generator: "@astrojs/netlify@${packageVersion}", path: "/*", preferStatic: true };
+			`
+		);
+	}
+
+	async function writeMiddleware(entrypoint: URL) {
+		await mkdir(middlewareOutputDir(), { recursive: true });
+		await writeFile(
+			new URL('./entry.mjs', middlewareOutputDir()),
+			`
+						import { onRequest } from "${fileURLToPath(entrypoint).replaceAll('\\', '/')}";
+						import { createContext, trySerializeLocals } from 'astro/middleware';
+
+						export default async (request, context) => {
+							const ctx = createContext({ 
+								request,
+								params: {}
+							});
+							ctx.locals = { netlify: { context } }
+							const next = () => {
+								const { netlify, ...otherLocals } = ctx.locals;
+								request.headers.set("x-astro-locals", trySerializeLocals(otherLocals));
```

`trySerializeLocals` can throw a parse error that we should catch.

---

What should we do about the error, if it occurs?

---

I think a `console.error` should be enough. What do you think? Will the user be able to see the stacktrace from their dashboard?

---

If we logged the error and then continued the request, it would be missing locals that might be important for authentication & authorization. So in the case of locals not being serializable, we definitely need to throw an exception and abort the request.
I'd say we keep it like it is - the user will be able to see the stacktrace from their dashboard already, and the error message thrown by `trySerializeLocals` should be enough to debug what's happening.